tapchannel/aux_sweeper.go

```diff
@@ -0,0 +1,1475 @@
+	)
+	if err != nil {
+		return lfn.Err[tapscriptSweepDesc](err)
+	}
+
+	// Now that we have the script tree, we'll make the control block
+	// needed to spend it, but taking the revoked path.
+	ctrlBlock, err := toLocalScriptTree.CtrlBlockForPath(
+		input.ScriptPathRevocation,
+	)
+	if err != nil {
+		return lfn.Err[tapscriptSweepDesc](err)
+	}
+	ctrlBlockBytes, err := ctrlBlock.ToBytes()
+	if err != nil {
+		return lfn.Err[tapscriptSweepDesc](err)
+	}
+
+	return lfn.Ok(tapscriptSweepDesc{
+		scriptTree:     toLocalScriptTree,
+		ctrlBlockBytes: ctrlBlockBytes,
+	})
+}
+
+// addVPacket converts an asset outputs to the corresponding vPackets that can
+// be used to complete the proof needed to import a commitment transaction.
+// This new vPacket is added to the specified map.
+func assetOutputToVPacket(fundingInputProofs map[asset.ID]*proof.Proof,
+	vPktsByAssetID map[asset.ID]*tappsbt.VPacket,
+	assetOutput *cmsg.AssetOutput, chainParams *address.ChainParams,
+	commitTx *wire.MsgTx, courierAddr *url.URL) error {
+
+	// Make a vPacket with the input set from the funding input proof for
+	// this asset ID.
+	var err error
+	assetID := assetOutput.AssetID.Val
+	vPkt, ok := vPktsByAssetID[assetID]
+	if !ok {
+		fundingInputProof, ok := fundingInputProofs[assetID]
+		if !ok {
+			return fmt.Errorf("funding input proof not found for "+
+				"asset ID: %v", assetOutput.AssetID)
+		}
+		vPkt, err = tappsbt.FromProofs(
+			[]*proof.Proof{fundingInputProof}, chainParams,
+		)
+		if err != nil {
+			return fmt.Errorf("unable to create "+
+				"vPacket: %v", err)
+		}
+
+		vPktsByAssetID[assetID] = vPkt
+	}
+
+	// With the input set, we'll now add the extra information for the
+	// output to the packet to complete it.
+	assetProof := assetOutput.Proof.Val
+	inclusionProof := assetProof.InclusionProof
+
+	scriptKey := asset.NewScriptKey(assetProof.Asset.ScriptKey.PubKey)
+	scriptKey.TweakedScriptKey = &asset.TweakedScriptKey{
+		RawKey: keychain.KeyDescriptor{
+			PubKey: assetProof.Asset.ScriptKey.PubKey,
+		},
+	}
+
+	vOut := &tappsbt.VOutput{
+		Amount:       assetProof.Asset.Amount,
+		AssetVersion: assetProof.Asset.Version,
+		Type: func() tappsbt.VOutputType {
+			if assetProof.Asset.SplitCommitmentRoot == nil {
+				return tappsbt.TypeSimple
+			}
+			return tappsbt.TypeSplitRoot
+		}(),
+		Interactive:             true,
+		AnchorOutputIndex:       inclusionProof.OutputIndex,
+		AnchorOutputInternalKey: inclusionProof.InternalKey,
+		//nolint:ll
+		AnchorOutputTapscriptSibling: inclusionProof.CommitmentProof.TapSiblingPreimage,
+		// TODO(roasbeef): want full info here actually?
+		ScriptKey:   scriptKey,
+		ProofSuffix: &assetProof,
+	}
+
+	// While we're here, we'll also replace the transaction stored in the
+	// proof with the correct one.
+	vOut.ProofSuffix.AnchorTx = *commitTx
+
+	// Finally, we'll set the delivery address to the default courier, so
+	// we publish the proof in the specified Universe.
+	vOut.ProofDeliveryAddress = courierAddr
```

Comment:
I don't think this comment is correct in general. The courier doesn't necessarily push to a universe server.

Reply:
Good point. This is in place mainly so we can run assertions properly in the itests. Could consider making the param an option, then it would only be done with some other config guard.